### PR TITLE
TYPE: fold the `r#` in raw-identifiers

### PR DESCRIPTION
**Pull request by ice1000**
Signed-off-by: ice1000 <ice1000kotlin@foxmail.com>

Before:
Screenshot from 2019-04-25 11-15-32
After:
Screenshot from 2019-04-25 11-15-13

### CODE REVIEW

**Comment by ice1000** [comment]
I did some general cleanup in related files.

**Comment by ice1000** [comment]
I have received some negative comments on this feature. Maybe I should make it disabled by default (but how can I test it then?).

**Comment by mchernyavsky** [comment]
> I have received some negative comments on this feature. Maybe I should make it disabled by default (but how can I test it then?).

I'm not sure, but it seems that in tests all regions are folded (regardless of the settings).

**Comment by ice1000** [comment]
@mchernyavsky I've modified the tests to let it work, please take a look.

File: src/main/kotlin/org/rust/ide/folding/RsFoldingBuilder.kt [hunk 1]
```diff
@@ -187,20 +198,21 @@ class RsFoldingBuilder : CustomFoldingBuilder(), DumbAware {
 
         private fun isInRangesAlready(ranges: MutableList<TextRange>, element: PsiElement?): Boolean {
             if (element == null) return false
-            return !ranges.filter { x -> x.contains(element.textOffset) }.isEmpty()
+            return ranges.any { x -> element.textOffset in x }
```
Comment (ice1000) [inline src/main/kotlin/org/rust/ide/folding/RsFoldingBuilder.kt]: I believe `any` will be faster than `isEmpty . filter` 😉

File: src/test/kotlin/org/rust/ide/folding/RsFoldingBuilderTest.kt [hunk 2]
```diff
@@ -10,7 +10,12 @@ import org.rust.RsTestBase
 class RsFoldingBuilderTest : RsTestBase() {
     override val dataPath = "org/rust/ide/folding/fixtures"
 
+    init {
+        RsCodeFoldingSettings.instance.hideRawKeywordsPrefix = true
```
Comment (ice1000) [inline src/test/kotlin/org/rust/ide/folding/RsFoldingBuilderTest.kt]: Enable by default

**Comment by ice1000** [comment]
bors ping

File: src/test/kotlin/org/rust/ide/folding/RsFoldingBuilderTest.kt (hunk 2 above)
Comment (mchernyavsky) [reply src/test/kotlin/org/rust/ide/folding/RsFoldingBuilderTest.kt]: Have you tried to run tests without this line? All tests pass for me even with `RsCodeFoldingSettings.instance.hideRawKeywordsPrefix = false`.

Also, after upgrading to the latest version of Kotlin plugin (1.3.31-release-IJ2019.1-1) I have this:

```
junit.framework.AssertionFailedError: Exception in constructor: test raw identifiers (java.lang.NullPointerException
	at com.intellij.openapi.components.ServiceManager.doGetService(ServiceManager.java:47)
	at com.intellij.openapi.components.ServiceManager.getService(ServiceManager.java:26)
	at org.rust.ide.folding.RsCodeFoldingSettings$Companion.getInstance(RsCodeFoldingSettings.kt:17)
	at org.rust.ide.folding.RsFoldingBuilderTest.<init>(RsFoldingBuilderTest.kt:14)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at junit.framework.TestSuite.createTest(TestSuite.java:60)
	at junit.framework.TestSuite.addTestMethod(TestSuite.java:307)
	at junit.framework.TestSuite.addTestsFromTestCase(TestSuite.java:150)
	at junit.framework.TestSuite.<init>(TestSuite.java:129)
	at org.junit.internal.runners.JUnit38ClassRunner.<init>(JUnit38ClassRunner.java:74)
	at org.junit.internal.builders.JUnit3Builder.runnerForClass(JUnit3Builder.java:11)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.requests.ClassRequest.getRunner(ClassRequest.java:33)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:49)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
)

	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.TestCase.fail(TestCase.java:227)
	at junit.framework.TestSuite$1.runTest(TestSuite.java:97)
	at junit.framework.TestCase.runBare(TestCase.java:141)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

If `hideRawKeywordsPrefix = true` is necessary, I would replace `init` with `setUp`:
```rust
override fun setUp() {
    super.setUp()
    RsCodeFoldingSettings.instance.hideRawKeywordsPrefix = true
}
```

Comment (ice1000) [reply src/test/kotlin/org/rust/ide/folding/RsFoldingBuilderTest.kt]: Fixed by removing this line. :wink:

**Comment by mchernyavsky** [comment]
LGTM, thanks!
bors r+

**Comment by Undin** [comment]
Hm, I'm not sure that it's a useful feature
Some reasons:
* raw identifiers are supposed to be used for keyword-like identifiers that quite rare case. And in this case, folding of `r#` can be misleading because folded item name conflicts with the corresponding keyword
* current implementation hides `r#` prefix completely. It can be confusing while modification of the identifier and navigation when the prefix appears suddenly. 


But other implementations don't make sense at all because `r#` prefix is already very short

Due to the described reasons, we cannot turn on it by default (in this PR is just turned off) and looks like won't do it in the future. Including settings complication, nobody won't discover it. 

Finally:
* usefulness is rather ambiguous 
* it's not discoverable because it's turned off by default

So I'm going to revert these changes

@ice1000 Sorry for so non-instant feedback

**Comment by ice1000** [comment]
Yeah that's what I heard from my friends too. No problem, that's reasonable.

**Pull request merged**